tick/stateful/stateful_expr.go

```diff
@@ -0,0 +1,95 @@
+package stateful
+
+import (
+	"fmt"
+
+	"github.com/influxdata/kapacitor/tick"
+)
+
+// StatefulExpression is interface that describe expression with state and
+// it's evaluation.
+type StatefulExpression interface {
+	Reset()
+
+	EvalFloat(scope *tick.Scope) (float64, error)
+	EvalInt(scope *tick.Scope) (int64, error)
+	EvalString(scope *tick.Scope) (string, error)
+	EvalBool(scope *tick.Scope) (bool, error)
+
+	EvalNum(scope *tick.Scope) (interface{}, error)
+}
+
+type StatefulExpr struct {
+	nodeEvaluator  NodeEvaluator
+	executionState ExecutionState
+}
+
+// TryCreateStatefulExpr accept a node and try to "compile"/ "specialise" it
+// in order to achieve better runtime performance.
+//
+// For example:
+// 	Given a BinaryNode{ReferNode("value"), NumberNode{Float64:10}} during runtime
+// 	we can find the type of "value" and find the most matching comparison function - (float64,float64) or (int64,float64)
+//
+// If the given node type isn't supported for compilation we will fallback to "NewStatefulExpr"
+func TryCreateStatefulExpr(node tick.Node) (StatefulExpression, error) {
```

Comment:
Can we name this `NewExpression`, since it returns an error we know its a "try" operation. And again we can drop the `Stateful` part in the name.